### PR TITLE
Fix SmtpClient reuse with SendAsync

### LINKED ISSUE #49340: Reusing SmtpClient hangs in net50
Calling `SendMailAsync` multiple times on the same `SmtpClient` instance fails under some circumstances in net50.

### Description

The following test succeeds when running under netcoreapp3.1, but fails by timing out in net50

```C#
[Fact(Timeout = 10000)]
public async Task Net50Repo()
{
    var config = Configuration.Configure();
    using var server = SimpleSmtpServer.Start(config.WithRandomPort().Port);
    var inst = new SmtpClient
    {
        Host = "localhost",
        Port = server.Configuration.Port,
        EnableSsl = false
    };

    for (int messageNo = 0; messageNo < 2; messageNo++)
    {
        var mailMessage = new MailMessage("one@example.com", "two@example.com");

        await inst.SendMailAsync(mailMessage);

        Assert.Equal(messageNo + 1, server.ReceivedEmailCount);
    }
}
```
The issue seems to be reusing the `SmtpClient`, the first `SendMailAsync` succeeds, but the second call always just hangs indefinitely.

Works in 3.1. Newing up a new client for each call also works. Testing against a different Smtp server (eg. Papercut) also works.

netDumpster just ends up waiting inside `SmtpContext` on this line:

`count = this.socket.Receive(byteBuffer);`
It's like if the socket remains open it cannot be reused anymore for some reason. Any idea what changed?

### Configuration

We are able to reproduce with netdumpster in unit tests, production is not yet running in net50 for us (partly because of this), so I'm unsure if other smtp server are effected, we are nervous about this.

### Regression?

Yes, regression from 3.1 to 5

### Other information

I originally raised the issue on [netdumpster](https://github.com/cmendible/netDumbster/issues/30), there is some more discussion there, although not much.

I am aware that other issues mention `SmtpClient` is not under active development, and it seems MS recommends migrating to MailKit, but it would still be nice if existing functionality would not break.

@cmendible @scalablecory

### PR DESCRIPTION
**Pull request by vonzshik**
Fixes #49340

Not entirely happy with how it turned out (with passing around `ContextAwareResult` as `object`). On the other hand, it already was like this, just hidden behind `IAsyncResult`...

### CODE REVIEW

File: src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs [hunk 1]
```diff
@@ -648,8 +648,23 @@ public void SendAsync(MailMessage message, object? userToken)
                         _operationCompletedResult = new ContextAwareResult(_transport.IdentityRequired, true, null, this, s_contextSafeCompleteCallback);
                         lock (_operationCompletedResult.StartPostingAsyncOp())
                         {
-                            if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"Calling BeginConnect. Transport: {_transport}");
-                            _transport.BeginGetConnection(_operationCompletedResult, ConnectCallback, _operationCompletedResult, Host!, Port);
+                            if (_transport.IsConnected)
+                            {
+                                try
+                                {
+                                    SendMailAsync(_operationCompletedResult);
```
Comment (wfurt) [inline src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs:655]: It worries me that we will now have more/different code under the lock.

File: src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs [hunk 2]
```diff
@@ -829,9 +844,9 @@ private void TimeOutCallback(object? state)
             }
         }
 
-        private void Complete(Exception? exception, IAsyncResult result)
+        private void Complete(Exception? exception, object result)
```
Comment (wfurt) [inline src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs:847]: what is motivation for this change? Using generic `object` is undesirable IMHO.

**Comment by wfurt** [comment]
any other suggestions @filipnavara ?

File: src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs (hunk 1 above)
Comment (vonzshik) [reply src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs:655]: It is what it is. The best way is to rewrite everything to Tasks, but with `SmtpClient` being deprecated, I can't think of anything better.

**Comment by karelz** [comment]
@filipnavara any thoughts?

**Comment by filipnavara** [comment]
Looks ok to me.

I don't think that rewriting the async logic is worth the trouble for an API that is marked obsolete.

**Review by wfurt — APPROVED** [review]
LGTM

**Pull request merged**